packages/core/http/core-http-versioned-router-server-internal/src/core_versioned_route.ts

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { schema } from '@kbn/config-schema';
+import type {
+  RequestHandler,
+  IRouter,
+  RequestHandlerContextBase,
+  KibanaRequest,
+  KibanaResponseFactory,
+} from '@kbn/core-http-server';
+import type {
+  Version,
+  AddVersionOpts,
+  VersionedRoute,
+  VersionedRouteConfig,
+} from '@kbn/core-http-server';
+import type { CoreKibanaRequest } from '@kbn/core-http-router-server-internal';
+import type { Mutable } from 'utility-types';
+import type { Method } from './types';
+
+import { validate } from './validate';
+
+type Options = AddVersionOpts<unknown, unknown, unknown, unknown>;
+
+/** @internal */
+export const VERSION_HEADER = 'Elastic-Api-Version';
+
+// This validation is a pass-through so that we can apply our version-specific validation later
+const passThroughValidation = { body: schema.any(), params: schema.any(), query: schema.any() };
+
+export class CoreVersionedRoute implements VersionedRoute {
+  private readonly handlers = new Map<
+    Version,
+    {
+      fn: RequestHandler;
+      options: Options;
+    }
+  >();
+
+  public static from({
+    router,
+    method,
+    path,
+    options,
+  }: {
+    router: IRouter;
+    method: Method;
+    path: string;
+    options: VersionedRouteConfig<Method>;
+  }) {
+    return new CoreVersionedRoute(router, method, path, options);
+  }
+
+  private constructor(
+    private readonly router: IRouter,
+    public readonly method: Method,
+    public readonly path: string,
+    public readonly options: VersionedRouteConfig<Method>,
+    // TODO: Make "true" dev-only
+    private readonly validateResponses: boolean = true
```

Comment:
Is it a `TODO` for now or a `TODO` for later 😄?

Reply:
I was planning on addressing this when exposing the versioned router on our http router instance :thu: